File: demo/bridge-demo-adr/app/src/main/java/com/viethai/demo/MediaModule.kt
package com.viethai.demo

import io.reactivex.Flowable
import java.util.concurrent.TimeUnit

/** Created by viethai.pham on 4/4/19 */
class MediaModule {
  sealed class PlayEvent(val type: String) {
    object StartVideo : PlayEvent("START_VIDEO")
    object StopVideo : PlayEvent("STOP_VIDEO")
    data class ElapsedTime(val elapsed: Int, val total: Int) : PlayEvent("PROGRESS_VIDEO")
  }

  fun playVideo(): Flowable<PlayEvent> {
    val totalTime = 5

    return Flowable.interval(0, 1, TimeUnit.SECONDS)
      .take(totalTime.toLong() + 1)
      .map { PlayEvent.ElapsedTime(it.toInt(), totalTime) as PlayEvent }
      .startWith(PlayEvent.StartVideo)
      .concatWith(Flowable.just(PlayEvent.StopVideo))
  }
}

File: demo/bridge-demo-ios/BridgeDemoIOS/MediaModule.swift
//
//  MediaModule.swift
//  BridgeDemoIOS
//
//  Created by Viethai Pham on 5/4/19.
//  Copyright © 2019 Viet Hai Pham. All rights reserved.
//

import Foundation
import RxSwift

public final class MediaModule {
  public enum PlayEvent: ResponseType {
    case startVideo
    case stopVideo
    case elapsedTime(Int, Int)
    
    private var type: String {
      switch self {
      case .startVideo: return "START_VIDEO"
      case .stopVideo: return "STOP_VIDEO"
      case .elapsedTime: return "ELAPSED_TIME"
      }
    }
    
    public func toDictionary() -> [String : Any?] {
      var dict: [String : Any] = ["type" : self.type]
      
      switch self {
      case .elapsedTime(let elapsed, let total):
        dict["elapsed"] = elapsed
        dict["total"] = total
        
      default:
        break
      }
      
      return dict
    }
  }
  
  public func playVideo() -> Observable<PlayEvent> {
    let totalTime = 5
    
    return Observable<Int>
      .interval(1, scheduler: MainScheduler.instance)
      .take(totalTime + 1)
      .map({.elapsedTime($0, totalTime)})
      .startWith(.startVideo)
      .concat(Observable.just(.stopVideo))
  }
}
